client/components/data/query-jetpack-modules/index.jsx

```diff
@@ -1,21 +1,23 @@
+import { useQueryClient } from '@tanstack/react-query';
 import PropTypes from 'prop-types';
 import { useEffect } from 'react';
 import { useDispatch } from 'react-redux';
 import { fetchModuleList } from 'calypso/state/jetpack/modules/actions';
 import isFetchingJetpackModules from 'calypso/state/selectors/is-fetching-jetpack-modules';
 
-const request = ( siteId ) => ( dispatch, getState ) => {
+const request = ( siteId, queryClient ) => ( dispatch, getState ) => {
 	if ( siteId && ! isFetchingJetpackModules( getState(), siteId ) ) {
-		dispatch( fetchModuleList( siteId ) );
+		dispatch( fetchModuleList( siteId, queryClient ) );
```

Comment:
This is a big red flag. A redux action creator should not need or care about RQ at all. There is a high chance we'll break the rule of keeping action creators pure functions, and that may lead to unexpected and difficult-to-catch and reproduce bugs.

Reply:
Using only Redux makes sense. We could use a query component that only requests the data once. Feel free to do it in a new PR!

In terms of "stale" data, you can use the old data layer (look for polling for prior art), or alternatively, have a separate field with a timestamp of the last requested time, and then if we're beyond a certain interval since that timestamp, trigger a re-request (dispatch the request action again).

Reply:
I second what Marin said. If I understand you correctly this is how I'd go about it:

1. Have a `QueryJetpackConnectionHealth` component that is part of the tree everywhere we need it
    - There's a bunch of examples in [`client/components/data`](https://github.com/Automattic/wp-calypso/tree/trunk/client/components/data). 
    - This component only issues a request, which can be in a Redux thunk or utilize the (legacy) Redux data layer, if it hasn't done so already. How you determine that can either be by data being present or a separate field (like `lastUpdatedAt` or so).
2. Create a Redux selector which retrieves the data from the store
3. If you need to invalidate the data dispatch that exact Redux action which fetches data. You have full access to the Redux `state` in `fetchModuleList` thunk which is passed as [second parameter of the thunk](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/jetpack/modules/actions.js#L137) `getState`.